### ISSUE TITLE
unregister() in service worker navigation fetch event

### ISSUE THREAD

**Issue by wanderview** [slightlyoff/ServiceWorker#804]
Consider this case:

```
addEventListener('fetch', function(evt) {
  fetch.respondWith(self.registration.unregister().then(function() {
    return new Response('foo');
  });
});
```

If this is the first navigation for this scope, then there may be no controlled clients.  This means the unregister() can kill the current worker before it completes the fetch event.

This leaves things in a bit of a messy state because .respondWith() was called, but the worker died before it completed.

This is kind of a corner case, but it would be nice to understand the desired behavior here.  Should we simply fall back to a normal network load?  Or should we keep the worker alive until the fetch event is complete?  Imagine if someone wanted to call register() after the unregister().

**Comment by jungkees** [comment]
> If this is the first navigation for this scope, then there may be no controlled clients.

Once a fetch event is dispatched to the service worker in the example, that means the client got a controller for this fetch already. If there was no matching registration or registration's active worker, it should have hit the network.

**Comment by wanderview** [comment]
> Once a fetch event is dispatched to the service worker in the example, that means the client got a controller for this fetch already. If there was no matching registration or registration's active worker, it should have hit the network.

I guess I am talking about this scenario:

1) https://example.com/index.html calls register('foo/sw.js') which results in a registration with scope https://example.com/foo/
2) https://example.com/foo/index.html is opened in a new tab
3) A FetchEvent for foo/index.html is fired

At point 3 there are no loaded documents under foo/ yet.

Are you referring to this test after 12.5 in Handle Fetch?

> Note: From this point, the service worker client starts to use its active worker's containing service worker registration.

Would the unloaded (about:blank?) document in the client at this point start returning the service worker for navigator.serviceWorker.controller even before its initial network load is handled by the service worker?

**Comment by jungkees** [comment]
> Are you referring to this text after 12.5 in Handle Fetch?

Yes.

> Would the unloaded (about:blank?) document in the client at this point start returning the service worker for navigator.serviceWorker.controller even before its initial network load is handled by the service worker?

I think the network load being handled by the service worker means the client initiated this request is already controlled that worker.

Also I wonder how one can access navigator.serviceWorker.controller from the unloaded (about:blank?) document when the navigation to a new document already occurred with [replacement enabled](https://html.spec.whatwg.org/multipage/browsers.html#replacement-enabled).

**Comment by wanderview** [comment]
Right, I guess I have a hard time reasoning about clients in this mid-navigation state.  Its weird for me to think of them as controlled.

Anyway, I think you've convinced me there is no spec issue here.

**Comment by wanderview** [comment]
I do have one question.  At what point does the client stop being controlled if the user navigates to a new page?  Or if the page is redirected to about:neterror or other internal error page, is the controlling state removed?

**Comment by jungkees** [comment]
We don't have a concept to reset the controller to null at the moment. I thought it was okay as a new client is created during a navigation. But I think it's not any more. It seems the behavior specified in HTML has been changed recently: https://html.spec.whatwg.org/#navigating-across-documents:set-up-a-browsing-context-environment-settings-object. Now an environment settings object is created only once per browsing context creation. This weakens the concept of a service worker client being an environment settings object. I think I have to dig into it.

**Comment by jungkees** [comment]
Will open a new issue for it.

**Comment by jungkees** [comment]
Filed: https://github.com/slightlyoff/ServiceWorker/issues/811
